File: July2022/Java/PartitionList.java

// @saorav21994

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode res = new ListNode(0);
        ListNode start = res;
        ListNode tmp = head;
        while (tmp != null) {
            if (tmp.val < x) {
                res.next = new ListNode(tmp.val);
                res = res.next;
            }
            tmp = tmp.next;
        }
        tmp = head;
        while (tmp != null) {
            if (tmp.val >= x) {
                res.next = new ListNode(tmp.val);
                res = res.next;
            }
            tmp = tmp.next;
        }
        return start.next;
    }
}

File: July2022/python/partitionList.py
@Jiganesh

# https://leetcode.com/problems/partition-list/

from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    # Runtime: 49 ms, faster than 61.47% of Python3 online submissions for Partition List.
    # Memory Usage: 14 MB, less than 31.57% of Python3 online submissions for Partition List.
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        
        smallerNode, largerNode = ListNode(-1), ListNode(-1)
        smallerPointer , largerPointer = smallerNode, largerNode
        current= head
        
        while current :
            
            nextNode = current.next
            current.next = None
            
            if current.val >=x :
                largerPointer.next = current
                largerPointer = largerPointer.next

            else:
                smallerPointer.next =current
                smallerPointer = smallerPointer.next

            current = nextNode
        
        smallerPointer.next = largerNode.next
        return smallerNode.next
            
